### PR TITLE
[lang] support regex option for parse operator

### PR DESCRIPTION
**Pull request by tstack**
This change adds support for `parse regex`.  It also fixes handling of escape sequences in quoted strings since they weren't really working before.

### CODE REVIEW

File: src/lang.rs [hunk 1]
```diff
@@ -1052,8 +1063,58 @@ fn req_ident(input: Span) -> IResult<Span, String> {
...
+fn escaped_chars(input: Span) -> IResult<Span, Span> {
+    alt((
+        tag("\\"),
+        tag("n"),
+        tag("n"),
+        tag("t"),
+        tag("0"),
+        tag("'"),
+        tag("\""),
+        take(1usize),
```
Comment (rcoh) [inline src/lang.rs:1075]: what does `take(1usize)` do here?

File: src/lang.rs [hunk 2]
```diff
@@ -1101,17 +1162,70 @@ fn var_list(input: Span) -> IResult<Span, Vec<String>> {
...
+            opt(with_pos(tag("as").preceded_by(multispace1).precedes(var_list))),
             opt(multispace1.precedes(with_pos(pair(tag("from"), multispace1).precedes(expr)))),
             opt(tag("nodrop").preceded_by(multispace1)).map(|nd| nd.is_some()),
         ))
-        .map(|(s, e1, fields, e2, no_drop)| InlineOperator::Parse {
-            pattern: Keyword::new_wildcard(s),
-            fields,
-            input_column: (e1, e2),
-            no_drop,
+        .map(|(_p, is_regex, s, e1, user_fields_opt, e2, no_drop)| {
```
Comment (rcoh) [inline src/lang.rs:1173]: mostly just curious: is there a way to have this be a struct instead of a tuple?

File: src/lang.rs [hunk 3]
```diff
@@ -1101,17 +1162,70 @@ fn var_list(input: Span) -> IResult<Span, Vec<String>> {
...
+                                .with_code_range(s.range, format!(
+                                    "contains {} unnamed capture(s)", unnamed_capture_count))
+                                .with_resolution("name the capture groups by adding -- ?P<capname> -- right after the left parenthesis")
+                                .send_report();
+                            Vec::new()
+                        } else if let Some(user_fields) = user_fields_opt {
+                            input
+                                .extra
+                                .report_error_for("the parse regex operator does not accept user-defined field names")
+                                .with_code_range(user_fields.range, "remove this")
```
Comment (rcoh) [inline src/lang.rs:1206]: I found this error message is a little confusing. What about something like:

"the parse regex operator does not support an `as` clause" or something like that.

It took me a bit to figure out what the user-defined field names actually meant.

The resolution could say something like: "remove the `as` clause. The named capture groups in the regular expression determine the output column"

File: src/lang.rs (hunk 2 above)
Comment (rcoh) [inline src/lang.rs:1173]: can you rename `e2` to `from_col` or something like that? this was a little hard to follow

**Review by rcoh — COMMENTED** [review]
just a couple small cleanups. Pretty wild this doesn't actually require changing the parse operator itself!

File: tests/structured_tests/parse_regex.toml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+query = """db-1 response | parse regex "in (?P<duration>\\\\d+)ms" """
```
Comment (rcoh) [inline tests/structured_tests/parse_regex.toml:1]: that is a lot of escaping! (But is it just toml that we're escaping here?) The docs seem to indicate that those escapes wouldn't be required.

Comment (tstack) [reply tests/structured_tests/parse_regex.toml:1]: It's escaping for toml.  Looks like I can change it to a literal string (single-quote instead of double)

File: src/lang.rs (hunk 1 above)
Comment (tstack) [reply src/lang.rs:1075]: It's to allow passing through things like `\d`.  Added a comment.

File: src/lang.rs (hunk 2 above)
Comment (tstack) [reply src/lang.rs:1173]: I'm not seeing anything right now.

Comment (tstack) [reply src/lang.rs]: Found this crate -- https://crates.io/crates/struple -- that does what you'd want... and, hey, it was written for exactly this purpose.  Should I import it and give it a go?

**Review by rcoh — APPROVED** [review]
